### PR TITLE
Enable Mono for using faster invoke stubs

### PR DESCRIPTION
**Pull request by steveharter**
Add support for Mono to emit faster invoke stubs for both the new invoke APIs added in https://github.com/dotnet/runtime/pull/88415 as well as standard invoke.

Replacement for closed issue https://github.com/dotnet/runtime/issues/88952

### CODE REVIEW

**Comment by EgorBo** [comment]
just curious - is it possible to avoid boxed `Nullable<>` in the first place? I guess it's not possible to create ones with plain C#

File: src/libraries/System.Private.CoreLib/src/System/Reflection/InvokerEmitUtil.cs [hunk 1]
```diff
@@ -196,6 +210,14 @@ public static unsafe InvokeFunc_RefArgs CreateInvokeDelegate_RefArgs(MethodBase
             return (InvokeFunc_RefArgs)dm.CreateDelegate(typeof(InvokeFunc_RefArgs), target: null);
         }
 
+        private static void EmitTrueNullableUnbox(ILGenerator il, Type parameterType)
+        {
+            // Unbox the true Nullable<T> created by reflection to a Nullable<T> without using
+            // OpCodes.Unbox since unboxing that is not necessarily a valid CLI operation.
+            il.Emit(OpCodes.Call, Methods.Object_GetRawData());
+            il.Emit(OpCodes.Ldobj, parameterType);
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/InvokerEmitUtil.cs:218]: We do all type safety checks in the non-dynamic managed code.

Can you use this sequence for all pointers and valuetypes? It would skipt the type check for all them, just like we are skipping the type check for nullable here.

Comment (steveharter) [reply src/libraries/System.Private.CoreLib/src/System/Reflection/InvokerEmitUtil.cs:218]: Hmm I applied those changes to the other args but I think there may be a problem in general when invoke is used to call a method on a `Nullable<T>` -- this is based on changing the other .Unbox used for the `this` pointer to the new helper method where I saw some test failures. I'll update based on what I find.

File: src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs [hunk 2]
```diff
@@ -1793,8 +1793,15 @@ internal override FieldInfo GetField(FieldInfo fromNoninstanciated)
             else if (IsPointer)
             {
                 Type? vtype = value.GetType();
-                if (vtype == typeof(IntPtr) || vtype == typeof(UIntPtr))
+                if (vtype == typeof(IntPtr))
                     return CheckValueStatus.Success;
+
```
Comment (vargaz) [inline src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:1798]: Why is this needed ?

Comment (steveharter) [reply src/mono/System.Private.CoreLib/src/System/RuntimeType.Mono.cs:1797]: Thanks for this. With the last commit, it is no longer necessary. Also removed from core CLR.

It was necessary previously since the `Opcodes.UnboxAny` would throw a CastClassException.

**Review by jkotas — APPROVED** [review]

File: src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs [hunk 3]
```diff
@@ -80,14 +80,14 @@ internal static void ThrowTargetParameterCountException()
             bool copyBack = false;
             Span<bool> shouldCopyBack = new(ref copyBack);
 
-            CheckArguments(parametersSpan, copyOfArgs, shouldCopyBack, binder, culture, invokeAttr);
```
Comment (steveharter) [inline src/libraries/System.Private.CoreLib/src/System/Reflection/MethodBaseInvoker.cs:83]: FYI this file contains misc, unrelated consistency changes.
